drivers: ieee802154: cc13xx_cc26xx: adopt hal/ti rf driverlib

cfriedt opened the pull request:
This change reworks the cc13xx_cc26xx IEEE 802.15.4 driver to use the TI RF driver API that is available in modules/hal/ti.

There are a number of benefits to using TI's API including
 - a stable multi-OS vendor library and API
 - API compatibility with the rest of the SimpleLink SDK and SoC family
 - potential multi-protocol & multi-client radio operation (e.g. both 15.4 and BLE)
 - coexistence support with other chipsets via gpio
 - vetted TI RF driver resources, such as
   - the radio command queue
   - highly tuned / coupled RTC & RAT (RAdio Timer) API

Fixes #26312
Fixes #29418

cfriedt commented:
Just a first commit so that it builds.

The `PowerCC26X2_Config` needs to be properly initialized - @vanti are you doing some power work still?

Need to hook up IRQs. I think that's done internally inside of the RF DriverLib, but I can't remember off the top of my head. 

Rather than using the doorbell api, we should use the RF API for sending commands -`RF_postCmd()`. 

I figure a reasonable test to make sure this is working is `echo_client` and `echo_server`.

cfriedt commented:
Builds on my machine without issue using
```
west build samples/net/sockets/echo_server -- -DCONF_FILE="prj.conf overlay-802154.conf"
```

vanti commented:
@cfriedt The `PowerCC26X2_config` structure is currently defined in Zephyr under `soc/arm/ti_simplelink/cc13x2_cc26x2/power.c`. I'd expect you to customize it to suit your needs. You would have to enable system power management in Zephyr in order for this structure to be linked in today:

```
CONFIG_SYS_POWER_MANAGEMENT=y
CONFIG_SYS_POWER_SLEEP_STATES=y
```

cfriedt commented:
> @cfriedt The `PowerCC26X2_config` structure is currently defined in Zephyr under `soc/arm/ti_simplelink/cc13x2_cc26x2/power.c`. I'd expect you to customize it to suit your needs. You would have to enable system power management in Zephyr in order for this structure to be linked in today:
> 
> ```
> CONFIG_SYS_POWER_MANAGEMENT=y
> CONFIG_SYS_POWER_SLEEP_STATES=y
> ```

Thanks fort the heads up - I will likely be asking you a few more questions on power management in the near future. I have to divert my attention to Greybus for the next little bit, but will be getting back to this PR shortly.

cfriedt commented:
I'm going to need to get back to this shortly.

cfriedt commented:
> @cfriedt The `PowerCC26X2_config` structure is currently defined in Zephyr under `soc/arm/ti_simplelink/cc13x2_cc26x2/power.c`. I'd expect you to customize it to suit your needs. You would have to enable system power management in Zephyr in order for this structure to be linked in today:
> 
> ```
> CONFIG_SYS_POWER_MANAGEMENT=y
> CONFIG_SYS_POWER_SLEEP_STATES=y
> ```

So far, I'll I've done is linked it in.

From what I remember, the RF Driver runs calibrateFxn regardless of whether or not it is NULL. I might need to bake a change inside of soc/arm/ti_simplelink/cc13x2_cc26x2/power.c since the `PowerCC26X2_Config` is const.

cfriedt commented on drivers/ieee802154/ieee802154_cc13xx_cc26xx.c:
```diff
@@ -225,9 +265,12 @@ static int ieee802154_cc13xx_cc26xx_tx(const struct device *dev,
 	__ASSERT_NO_MSG(k_sem_count_get(&drv_data->fg_done) == 0);
 
 	do {
```
@bwitherspoon @vanti - I wonder if it would be possible to convert the TX ISR to a SWI callback as well, sort of like what I did for the RX ISR. The RF core supports a number of configurable conditions.. I wonder if it can count down too..

Edit:  Aha!
https://www.ti.com/lit/pdf/swcu185
25.3.3.1.11 CMD_COUNT: Counter Command

cfriedt replied:
> 25.3.3.1.11 CMD_COUNT: Counter Command

Actually
25.3.3.1.13 CMD_COUNT_BRANCH
might be more appropriate. Also
25.3.3.1.14 CMD_PATTERN_CHECK.

It looks as though the `RF_runCmd()` call could be used to run a chain of commands of the form:

CSMA -> IEEE_TX -> IEEE_RX_ACK -> PATTERN_CHECK -> CMD_COUNT_BRANCH
or
CSMA -> IEEE_TX -> PATTERN_CHECK -> CMD_COUNT_BRANCH

based on whether an ACK was requested. The PATTERN_CHECK pattern would check to see if IEEE_TX had been successful.

The pNextOp for PATTERN_CHECK would be NULL if the IEEE_TX were successful. Otherwise, we would get to the COUNT_BRANCH command.

The pNextOp for COUNT_BRANCH would be IEEE_TX again if the count were non-zero.

Should be interesting to see if there is a performance improvement.

cfriedt replied:
The radio command graph I was thinking for this would be something along these lines:
```graphviz
digraph G {

  start -> CSMA;
  CSMA -> "PATTERN_CHECK\nCSMA OK?";
  "PATTERN_CHECK\nCSMA OK?" -> CSMA [label = "no"];
  "PATTERN_CHECK\nCSMA OK?" -> IEEE_TX [label = "yes"];
  IEEE_TX -> "PATTERN_CHECK\nIEEE_TX ok?";
  "PATTERN_CHECK\nIEEE_TX ok?" -> CSMA [label = "no"];
  "PATTERN_CHECK\nIEEE_TX ok?" -> IEEE_RX_ACK [label = "yes"];
  IEEE_RX_ACK -> "PATTERN_CHECK\nIEEE_RX_ACK ok?";
  "PATTERN_CHECK\nIEEE_RX_ACK ok?" -> "COUNT_BRANCH\nRADIO_TX_RETRIES" [label = "no"];
  "PATTERN_CHECK\nIEEE_RX_ACK ok?" -> end [label = "yes"];
  "COUNT_BRANCH\nRADIO_TX_RETRIES" -> CSMA [label = "> 0"];
  "COUNT_BRANCH\nRADIO_TX_RETRIES" -> end [label = "== 0"];
  
  CSMA[shape="rectangle"];
  "PATTERN_CHECK\nCSMA OK?"[shape="diamond"];
  IEEE_TX[shape="rectangle"];
  "PATTERN_CHECK\nIEEE_TX ok?"[shape="diamond"];
  IEEE_RX_ACK[shape="rectangle"];
  "PATTERN_CHECK\nIEEE_RX_ACK ok?"[shape="diamond"];
  "COUNT_BRANCH\nRADIO_TX_RETRIES"[shape="diamond"];
}
```
https://dreampuf.github.io/GraphvizOnline

cfriedt replied:
The radio command graph should look something like this:

```dot
digraph G {

  start -> CSMA;
  CSMA -> "PATTERN_CHECK\nCSMA OK?";
  "PATTERN_CHECK\nCSMA OK?" -> "COUNT_BRANCH\nRADIO_TX_RETRIES" [label = "no"];
  "PATTERN_CHECK\nCSMA OK?" -> IEEE_TX [label = "yes"];
  IEEE_TX -> "PATTERN_CHECK\nIEEE_TX ok?";
  "PATTERN_CHECK\nIEEE_TX ok?" -> "COUNT_BRANCH\nRADIO_TX_RETRIES" [label = "no"];
  "PATTERN_CHECK\nIEEE_TX ok?" -> IEEE_RX_ACK [label = "yes"];
  IEEE_RX_ACK -> "PATTERN_CHECK\nIEEE_RX_ACK ok?";
  "PATTERN_CHECK\nIEEE_RX_ACK ok?" -> "COUNT_BRANCH\nRADIO_TX_RETRIES" [label = "no"];
  "PATTERN_CHECK\nIEEE_RX_ACK ok?" -> end [label = "yes"];
  "COUNT_BRANCH\nRADIO_TX_RETRIES" -> CSMA [label = "> 0"];
  "COUNT_BRANCH\nRADIO_TX_RETRIES" -> end [label = "== 0"];
  
  CSMA[shape="rectangle"];
  "PATTERN_CHECK\nCSMA OK?"[shape="diamond"];
  IEEE_TX[shape="rectangle"];
  "PATTERN_CHECK\nIEEE_TX ok?"[shape="diamond"];
  IEEE_RX_ACK[shape="rectangle"];
  "PATTERN_CHECK\nIEEE_RX_ACK ok?"[shape="diamond"];
  "COUNT_BRANCH\nRADIO_TX_RETRIES"[shape="diamond"];
}
```
https://dreampuf.github.io/GraphvizOnline

cfriedt commented on soc/arm/ti_simplelink/cc13x2_cc26x2/power.c:
```diff
@@ -23,16 +23,28 @@
+/* This will need to include BLE eventually */
+	.policyInitFxn      = NULL,
+	.policyFxn          = &PowerCC26XX_doWFI,
+/* PowerCC26XX_standbyPolicy() is not yet available in hal/ti */
```
IIRC the .policyFxn is executed without a NULL check somewhere in the RF driver. I wonder if it would be possible to make a "no op" replacement call and, if so, whether that would effectively allow Zephyr to maintain power management control.

vanti replied:
Would you know where the call site is? I searched through source/ti/drivers/rf/RFCC26X2_multiMode.c in the HAL and did not find direct calls to the policyFxn. It's odd that the RF driver would call the power management policy function.

cfriedt replied:
Ah - you're right. I'll mark correct it to NULL. It was the calibrate function that gets executed without a NULL check.

cfriedt commented on drivers/ieee802154/ieee802154_cc13xx_cc26xx.c:
```diff
@@ -85,30 +151,29 @@ static int ieee802154_cc13xx_cc26xx_cca(const struct device *dev)
-		LOG_ERR("Failed to clear RX queue (0x%x)", status);
-		return -EIO;
-	}
+	RF_runImmediateCmd(drv_data->rf_handle,
```
Yes - actually, I found that there was an issue - not here, but with the call above. Apparently, CMD_ABORT is not a direct command but is actually an immediate command. It's fixed in my branch that adds SubGHz support, but not yet in this branch. Let me update.

cfriedt replied:
Just going to commit changes for SubGHz support and then I'll get back to this branch. Just pasting here for the time being.
```cpp
	/* Abort FG and BG processes */
	status = RF_runImmediateCmd(drv_data->rf_handle, (uint32_t *)&cmd_abort);
	if (status != RF_StatCmdDoneSuccess && status != RF_StatSuccess) {
		LOG_ERR("Failed to abort radio operations (%d)", status);
		return -EIO;
	}

	/* Set all RX entries to empty */
	status = RF_runImmediateCmd(drv_data->rf_handle,
		(uint32_t *)&drv_data->cmd_clear_rx);
	if (status != RF_StatCmdDoneSuccess && status != RF_StatSuccess) {
		LOG_ERR("Failed to clear RX queue (%d)", status);
		return -EIO;
	}
```

cfriedt commented:
Booo .. PR is red because I'm using `CONFIG_IEEE802154_CC13XX_CC26XX_SUB_GHZ`. I just got SubG working, but I guess I can wait until I make that PR to fix it.

```
Found references to undefined Kconfig symbols. If any of these are false
positives, then add them to UNDEF_KCONFIG_WHITELIST in check_compliance.py in the ci-tools repo.

If the reference is for a comment like /* CONFIG_FOO_* */ (or
/* CONFIG_FOO_*_... */), then please use exactly that form (with the '*'). The
CI check knows not to flag it.

More generally, a reference followed by $, @, {, *, or ## will never be
flagged.

CONFIG_IEEE802154_CC13XX_CC26XX_SUB_GHZ    soc/arm/ti_simplelink/cc13x2_cc26x2/power.c:28
##[error]Process completed with exit code 1.
```

cfriedt commented:
So, let's see a show of hands. 

Who would like to see this logic implemented in the RF core?

![tx-call-graph](https://user-images.githubusercontent.com/983494/93635050-9c071100-f9bf-11ea-9f21-11da6ca3cbbe.png)

The advantage is that we eliminate the `fg_done` semaphore so that TX logic is not bouncing between the M0 (RF core) and the M4 (System core). Putting it in the RF core is just a bit of extra legwork, but if people really want that feature here, I could probably add it to this PR. Otherwise, I'm fine leaving it for a later optimization.

@bwitherspoon
@jukkar
@vanti
@tbursztyka

cfriedt commented:
> The advantage is that we eliminate the `fg_done` semaphore so that TX logic is not bouncing between the M0 (RF core) and the M4 (System core). Eliminating that would probably make TX about as performant as it can get, and also give more cycles to the system core.
> 
> Putting it in the RF core is just a bit of extra legwork, but if people really want that feature here, I could probably add it to this PR. Otherwise, I'm fine leaving it for a later optimization.

I just tried chaining CMD_IEEE_CSMA with CMD_PATTERN_CHECK commands (when there were actually no background commands running at all), and unfortunately it does not seem to be possible - at least not in IEEE mode.

```
#define ERROR_WRONG_BG   0x0806   ///< FG level command not compatible with running BG level command
```

cfriedt commented:
One last change - I was able to remote the fg_done semaphore by switching to `RF_runScheduleCmd()` for foreground commands.

vanti commented:
@cfriedt I rebased your work onto the latest master branch to try the echo_server/client examples, and saw some data mismatch messages that I did not see with the original driver:

```
uart:~$ *** Booting Zephyr OS build zephyr-v2.4.0-63-gd8ec10689b23  ***
[00:00:00.011,657] <inf> net_config: Initializing network
[00:00:00.112,792] <inf> net_config: IPv6 address: 2001:db8::2
[00:00:00.112,915] <inf> net_echo_client_sample: Run echo client
[00:00:00.112,945] <inf> net_echo_client_sample: Network connected
[00:00:00.112,976] <inf> net_echo_client_sample: Starting...
[00:00:00.113,708] <dbg> net_echo_client_sample.send_udp_data: IPv6 UDP: Sent 81                                                                                                                                                              bytes
[00:00:10.113,830] <err> net_echo_client_sample: UDP IPv6: Data packet not recei                                                                                                                                                             ved
[00:00:10.114,196] <dbg> net_echo_client_sample.send_udp_data: IPv6 UDP: Sent 67                                                                                                                                                              bytes
[00:00:20.114,318] <err> net_echo_client_sample: UDP IPv6: Data packet not recei                                                                                                                                                             ved
[00:00:20.114,776] <dbg> net_echo_client_sample.send_udp_data: IPv6 UDP: Sent 78                                                                                                                                                              bytes
[00:00:30.114,898] <err> net_echo_client_sample: UDP IPv6: Data packet not recei                                                                                                                                                             ved
[00:00:30.115,173] <dbg> net_echo_client_sample.send_udp_data: IPv6 UDP: Sent 25                                                                                                                                                              bytes
[00:00:30.183,715] <dbg> net_echo_client_sample.process_udp_proto: IPv6 UDP: Rec                                                                                                                                                             eived and compared 25 bytes, all ok
[00:00:30.184,539] <dbg> net_echo_client_sample.send_udp_data: IPv6 UDP: Sent 98                                                                                                                                                              bytes
[00:00:30.193,206] <err> net_echo_client_sample: Invalid amount of data received                                                                                                                                                             : UDP IPv6
[00:00:30.193,206] <wrn> net_echo_client_sample: IPv6 UDP: Received and compared                                                                                                                                                              25 bytes, data mismatch
[00:00:40.184,631] <err> net_echo_client_sample: UDP IPv6: Data packet not recei                                                                                                                                                             ved
[00:00:40.184,997] <dbg> net_echo_client_sample.send_udp_data: IPv6 UDP: Sent 12                                                                                                                                                              bytes
[00:00:40.201,507] <dbg> net_echo_client_sample.process_udp_proto: IPv6 UDP: Rec                                                                                                                                                             eived and compared 12 bytes, all ok
[00:00:40.202,087] <dbg> net_echo_client_sample.send_udp_data: IPv6 UDP: Sent 67                                                                                                                                                              bytes
[00:00:40.209,625] <err> net_echo_client_sample: Invalid amount of data received                                                                                                                                                             : UDP IPv6
[00:00:40.209,625] <wrn> net_echo_client_sample: IPv6 UDP: Received and compared                                                                                                                                                              12 bytes, data mismatch
[00:00:40.224,304] <err> net_echo_client_sample: Invalid amount of data received                                                                                                                                                             : UDP IPv6
[00:00:40.224,304] <wrn> net_echo_client_sample: IPv6 UDP: Received and compared                                                                                                                                                              12 bytes, data mismatch
[00:00:40.238,891] <err> net_echo_client_sample: Invalid amount of data received                                                                                                                                                             : UDP IPv6
[00:00:40.238,891] <wrn> net_echo_client_sample: IPv6 UDP: Received and compared                                                                                                                                                              12 bytes, data mismatch
[00:00:50.202,178] <err> net_echo_client_sample: UDP IPv6: Data packet not received
[00:00:50.202,514] <dbg> net_echo_client_sample.send_udp_data: IPv6 UDP: Sent 30 bytes
[00:00:50.215,087] <dbg> net_echo_client_sample.process_udp_proto: IPv6 UDP: Received and compared 30 bytes, all ok
[00:00:50.215,454] <dbg> net_echo_client_sample.send_udp_data: IPv6 UDP: Sent 90 bytes
[00:00:50.233,093] <err> net_echo_client_sample: Invalid amount of data received: UDP IPv6
[00:00:50.233,093] <wrn> net_echo_client_sample: IPv6 UDP: Received and compared 30 bytes, data mismatch
[00:00:50.240,264] <err> net_echo_client_sample: Invalid amount of data received: UDP IPv6
[00:00:50.240,264] <wrn> net_echo_client_sample: IPv6 UDP: Received and compared 30 bytes, data mismatch
```

Are you seeing the same? It feels like something is amiss.

cfriedt commented:
@vanti - I just rebased on master (c02eb30bbfbe2487753eb52078015f5a617a3dd1) as well and am also the same thing.

Also, `CONFIG_IEEE802154=y` seems to be necessary now, otherwise no network interface comes up.

```
diff --git a/samples/net/sockets/echo_client/overlay-qemu_802154.conf b/samples/net/sockets/echo_client/overlay-qemu_802154.conf
index 2d33639485..44fc495f4d 100644
--- a/samples/net/sockets/echo_client/overlay-qemu_802154.conf
+++ b/samples/net/sockets/echo_client/overlay-qemu_802154.conf
@@ -15,6 +15,7 @@ CONFIG_NET_CONFIG_NEED_IPV4=n
 CONFIG_NET_CONFIG_MY_IPV4_ADDR=""
 CONFIG_NET_CONFIG_PEER_IPV4_ADDR=""
 
+CONFIG_IEEE802154=y
 CONFIG_NET_L2_IEEE802154=y
 CONFIG_NET_L2_IEEE802154_SHELL=y
 CONFIG_NET_L2_IEEE802154_LOG_LEVEL_INF=y
diff --git a/samples/net/sockets/echo_server/overlay-802154.conf b/samples/net/sockets/echo_server/overlay-802154.conf
index fa9e680027..23b56d4f3d 100644
--- a/samples/net/sockets/echo_server/overlay-802154.conf
+++ b/samples/net/sockets/echo_server/overlay-802154.conf
@@ -9,6 +9,7 @@ CONFIG_NET_CONFIG_NEED_IPV4=n
 CONFIG_NET_CONFIG_MY_IPV4_ADDR=""
 CONFIG_NET_CONFIG_PEER_IPV4_ADDR=""
 
+CONFIG_IEEE802154=y
 CONFIG_NET_L2_IEEE802154=y
 CONFIG_NET_L2_IEEE802154_SHELL=y
 CONFIG_NET_L2_IEEE802154_LOG_LEVEL_INF=y
```

cfriedt commented:
@vanti I'll go back and revisit this. I was mainly testing with TCP, but possibly because there is no inherent rate-limiting in UDP, I didn't catch something. Let me track it down in the next little while.

cfriedt commented:
It was a simple fix - good find!!

```diff
diff --git a/drivers/ieee802154/ieee802154_cc13xx_cc26xx.c b/drivers/ieee802154/ieee802154_cc13xx_cc26xx.c
index 8f68b701d8..04ea717b34 100644
--- a/drivers/ieee802154/ieee802154_cc13xx_cc26xx.c
+++ b/drivers/ieee802154/ieee802154_cc13xx_cc26xx.c
@@ -274,7 +274,7 @@ static int ieee802154_cc13xx_cc26xx_tx(const struct device *dev,
                reason = RF_runScheduleCmd(drv_data->rf_handle,
                        (RF_Op *)&drv_data->cmd_ieee_csma, &sched_params,
                        cmd_ieee_csma_callback, RF_EventLastFGCmdDone);
-               if ((reason & RF_EventLastCmdDone) == 0) {
+               if ((reason & RF_EventLastFGCmdDone) == 0) {
                        LOG_DBG("Failed to run command (0x%" PRIx64 ")",
                                reason);
                        continue;
```

cfriedt commented:
@vanti - I discovered what was stopping the show today.

The RF synth just goes to sleep after a certain amount of time. I have a workaround in my SubGHz branch for detecting and correcting this on the TX side. However, it seems to happen on the RX side as well.

I'll contact TI to see if there is possibly some way that we can subscribe to a notification or interrupt saying that the synth has powered-down.

cfriedt commented:
> @vanti - I discovered what was stopping the show today.

Correction - *today*. I confirmed root cause. 

There is a bug in the SimpleLink API (at least version 4.10, which is what we are using in Zephyr). I have not checked to see if it's fixed in the current version (4.30, iirc).

I can produce an ugly-ish workaround for Zephyr, but ultimately it's up to TI if they would like to use my fix for the SimpleLink SDK / RF driver.

cfriedt committed:
drivers: ieee802154: cc13xx_cc26xx: use ti rf driver api

This change reworks the cc13xx_cc26xx IEEE 802.15.4 driver to use
the TI RF driver API that is available in modules/hal/ti.

There are a number of benefits to using TI's API including
 - a stable multi-OS vendor library and API
 - API compatibility with the rest of the SimpleLink SDK and SoC family
 - potential multi-protocol & multi-client radio operation
   (e.g. both 15.4 and BLE)
 - coexistence support with other chipsets via gpio
 - vetted TI RF driver resources, such as
   - the radio command queue
   - highly tuned / coupled RTC & RAT (RAdio Timer) API

Fixes #26312

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt commented on soc/arm/ti_simplelink/cc13x2_cc26x2/power.c:
```diff
@@ -23,16 +23,27 @@
+	.policyInitFxn      = NULL,
+	.policyFxn          = NULL,
+	.calibrateFxn       = &PowerCC26XX_calibrate,
+	.enablePolicy       = true,
```
I'll disable it shortly. The only thing that needs to be populated is `calibrateFxn`, and it's probably safe to say that calibrating is a good idea (although, I'm open to using the `PowerCC26XX_noCalibrate()`.

cfriedt commented:
Like the energizer bunny - 6 hrs later ;-)
```
[06:08:18.272,064] <dbg> net_echo_client_sample.send_udp_data: IPv6 UDP: Sent 22 bytes
[06:08:18.293,334] <dbg> net_echo_client_sample.process_udp_proto: IPv6 UDP: Received and compared 22 bytes, all ok
[06:08:18.293,792] <dbg> net_echo_client_sample.send_udp_data: IPv6 UDP: Sent 1 bytes
[06:08:18.313,995] <dbg> net_echo_client_sample.process_udp_proto: IPv6 UDP: Received and compared 1 bytes, all ok
[06:08:18.314,636] <dbg> net_echo_client_sample.send_udp_data: IPv6 UDP: Sent 65 bytes
[06:08:18.343,841] <dbg> net_echo_client_sample.process_udp_proto: IPv6 UDP: Received and compared 65 bytes, all ok
[06:08:18.344,390] <dbg> net_echo_client_sample.send_udp_data: IPv6 UDP: Sent 27 bytes
[06:08:18.362,823] <dbg> net_echo_client_sample.process_udp_proto: IPv6 UDP: Received and compared 27 bytes, all ok
[06:08:18.363,403] <dbg> net_echo_client_sample.send_udp_data: IPv6 UDP: Sent 102 bytes
[06:08:18.399,017] <dbg> net_echo_client_sample.process_udp_proto: IPv6 UDP: Received and compared 102 bytes, all ok
[06:08:18.399,597] <dbg> net_echo_client_sample.send_udp_data: IPv6 UDP: Sent 100 bytes
[06:08:18.432,342] <dbg> net_echo_client_sample.process_udp_proto: IPv6 UDP: Received and compared 100 bytes, all ok
[06:08:18.432,830] <dbg> net_echo_client_sample.send_udp_data: IPv6 UDP: Sent 76 bytes
[06:08:18.474,426] <dbg> net_echo_client_sample.process_udp_proto: IPv6 UDP: Received and compared 76 bytes, all ok
[06:08:18.474,914] <dbg> net_echo_client_sample.send_udp_data: IPv6 UDP: Sent 4 bytes
[06:08:18.488,739] <dbg> net_echo_client_sample.process_udp_proto: IPv6 UDP: Received and compared 4 bytes, all ok
```

vanti reviewed (changes requested):
I am seeing some build warnings:

```
[113/209] Building C object zephyr/CMa...e802154/ieee802154_cc13xx_cc26xx.c.obj
/home/vwan/zephyrproject-work/zephyr/drivers/ieee802154/ieee802154_cc13xx_cc26xx.c: In function 'cmd_ieee_csma_callback':
/home/vwan/zephyrproject-work/zephyr/drivers/ieee802154/ieee802154_cc13xx_cc26xx.c:82:29: warning: passing argument 2 of 'update_saved_cmdhandle' discards 'volatile' qualifier from pointer target type [-Wdiscarded-qualifiers]
   82 |  update_saved_cmdhandle(ch, &drv_data->saved_cmdhandle);
      |                             ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/vwan/zephyrproject-work/zephyr/drivers/ieee802154/ieee802154_cc13xx_cc26xx.c:69:67: note: expected 'RF_CmdHandle *' {aka 'short int *'} but argument is of type 'volatile int16_t *' {aka 'volatile short int *'}
   69 | static void update_saved_cmdhandle(RF_CmdHandle ch, RF_CmdHandle *saved)
      |                                                     ~~~~~~~~~~~~~~^~~~~
/home/vwan/zephyrproject-work/zephyr/drivers/ieee802154/ieee802154_cc13xx_cc26xx.c: In function 'cmd_ieee_rx_callback':
/home/vwan/zephyrproject-work/zephyr/drivers/ieee802154/ieee802154_cc13xx_cc26xx.c:98:29: warning: passing argument 2 of 'update_saved_cmdhandle' discards 'volatile' qualifier from pointer target type [-Wdiscarded-qualifiers]
   98 |  update_saved_cmdhandle(ch, &drv_data->saved_cmdhandle);
      |                             ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/vwan/zephyrproject-work/zephyr/drivers/ieee802154/ieee802154_cc13xx_cc26xx.c:69:67: note: expected 'RF_CmdHandle *' {aka 'short int *'} but argument is of type 'volatile int16_t *' {aka 'volatile short int *'}
   69 | static void update_saved_cmdhandle(RF_CmdHandle ch, RF_CmdHandle *saved)
      |                                                     ~~~~~~~~~~~~~~^~~~~
```

Otherwise I can see the example run overnight. Looking good!

cfriedt commented on drivers/ieee802154/ieee802154_cc13xx_cc26xx.c:
```diff
@@ -268,6 +296,14 @@ static int ieee802154_cc13xx_cc26xx_tx(const struct device *dev,
 
 	k_mutex_lock(&drv_data->tx_mutex, K_FOREVER);
 
+	if (drv_data->saved_cmdhandle >= BIT(12) - 5) {
```
Good call!

cfriedt committed:
ieee802154: cc13xx_cc26xx: workaround for issue in ti driver

This is a temporary workaround for an issue in TI's RF Driver
API. A subsequent release of the SimpleLink SDK will mitigate
the need for it and it can be reverted when hal/ti receives
that update.

Fixes #29418

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

jukkar commented:
ping @bwitherspoon are you ok with this change?
